### Task-2 Model Devlopment 

####  Befoe each Model i did SMOTE because the class is highly immbalanced. 
## Data Preprocessing

### Class Imbalance
Initially, the dataset exhibited significant class imbalance, with the distribution of classes as follows:

| Class       | Count    |
|-------------|----------|
| Fraud       | 12268      |
| Not Fraud   | 116878  |

This imbalance posed challenges for model training, as machine learning algorithms may tend to favor the majority class.

### Applying SMOTE
To address this issue, we applied SMOTE for only training datasets, which generates synthetic examples of the minority class (fraud) to create a more balanced dataset. After applying SMOTE, the new class distribution was as follows:

| Class       | Count    |
|-------------|----------|
| Fraud       | 187004  |
| Not Fraud   | 187004  |

With the application of SMOTE, the dataset now contains an equal number of samples for both classes, allowing for a more effective model training process.


I applied six different models to the balanced dataset, logging key metrics to monitor performance. The models included:

1. **Logistic Regression**
2. **Decision Tree Classifier**
3. **Random Forest Classifier**
4. **Gradient Boosting Classifier**
5. **Multi-Layer Perceptron (MLP)**
6. **LSTM and RNN Models and RNN** 

In [1]:
import sys 
import pandas as pd
sys.path.insert(0, '/10 A KAI 2/WEEK 8&9/Fraud_detection/')

In [2]:
from script.model1 import FraudModel

In [3]:
Data = pd.read_csv('../Data/merged.csv', index_col=0)

In [4]:
Data.isnull().sum()

user_id                              0
signup_time                          0
purchase_time                        0
purchase_value                       0
device_id                            0
age                                  0
ip_address                           0
class                                0
lower_bound_ip_address               0
upper_bound_ip_address               0
country                              0
purchase_day_of_week                 0
signup_day_of_week                   0
signup_hour                          0
purchase__hour                       0
time_diff                       148848
transaction_count                    0
user_transaction_frequency           0
device_transaction_frequency         0
user_transaction_velocity       148848
device_transaction_velocity     148848
category_Ads                         0
category_Direct                      0
category_SEO                         0
category_Chrome                      0
category_FireFox         

In [5]:
pd.set_option('display.max_columns', None)

In [6]:
Data['class'].value_counts()

class
0    134922
1     13926
Name: count, dtype: int64

In [7]:
Data = Data.drop(columns=['lower_bound_ip_address','upper_bound_ip_address','time_diff', 'transaction_count','signup_time','purchase_time' ,'user_transaction_frequency','device_transaction_frequency', 'user_transaction_velocity','device_transaction_velocity','device_id','ip_address', 'device_id_encoded'], axis=1)

In [8]:
model = FraudModel(Data, dataset_name="Fraud", source="Fraud Data Source")

In [9]:
Data

user_id  purchase_value       age  class        country  \
107357        2        0.932499 -0.944593      0  United States   
46439         4        0.222645  0.564192      0      Singapore   
88685         8        0.550270 -0.944593      0  United States   
24733        12       -0.104980 -1.640956      0      Singapore   
96823        16       -1.524687 -0.132170      0  United States   
...         ...             ...       ...    ...            ...   
130664   399992        2.461414  3.001461      0         France   
27356    399993       -0.924042  0.448132      0      Singapore   
55411    399995       -0.214188  0.216011      0      Singapore   
52841    399997        1.587748  1.724797      0      Singapore   
118474   400000        1.150915 -0.944593      1  United States   

        purchase_day_of_week  signup_day_of_week  signup_hour  purchase__hour  \
107357              0.991051            1.480767    -1.228507       -0.220243   
46439               0.991051           -1.010572     0.651603        1.370816   
88685              -0.006049           -0.014036    -0.650012       -0.075601   
24733              -0.504599            0.982500    -0.794635        1.226174   
96823              -0.006049           -1.010572     0.217732        0.069040   
...                      ...                 ...          ...             ...   
130664             -0.504599           -0.014036     0.796227       -0.654169   
27356              -0.006049           -0.014036    -1.228507       -0.654169   
55411               0.492501           -1.010572    -0.939259        1.660100   
52841              -0.504599           -1.010572    -0.360764        1.226174   
118474             -1.501699           -1.508839     1.663970       -0.798810   

        category_Ads  category_Direct  category_SEO  category_Chrome  \
107357         False            False          True             True   
46439          False             True         False            False   
88685          False            False          True             True   
24733           True            False         False            False   
96823          False             True         False            False   
...              ...              ...           ...              ...   
130664         False             True         False            False   
27356          False             True         False            False   
55411          False             True         False            False   
52841          False             True         False             True   
118474          True            False         False             True   

        category_FireFox  category_IE  category_Opera  category_Safari  \
107357             False        False           False            False   
46439              False        False           False             True   
88685              False        False           False            False   
24733              False        False           False             True   
96823              False         True           False            False   
...                  ...          ...             ...              ...   
130664             False        False           False             True   
27356              False         True           False            False   
55411               True        False           False            False   
52841              False        False           False            False   
118474             False        False           False            False   

        category_F  category_M  Country_encoded  
107357        True       False         0.083790  
46439         True       False        -0.036750  
88685        False        True         0.083790  
24733        False        True        -0.036750  
96823        False        True         0.083790  
...            ...         ...              ...  
130664       False        True        -0.501673  
27356         True       False        -0.036750  
55411         True       F

In [10]:
model.fraud_preprocess()

purchase_value       age  class  purchase_day_of_week  \
107357        0.932499 -0.944593      0              0.991051   
46439         0.222645  0.564192      0              0.991051   
88685         0.550270 -0.944593      0             -0.006049   
24733        -0.104980 -1.640956      0             -0.504599   
96823        -1.524687 -0.132170      0             -0.006049   
...                ...       ...    ...                   ...   
130664        2.461414  3.001461      0             -0.504599   
27356        -0.924042  0.448132      0             -0.006049   
55411        -0.214188  0.216011      0              0.492501   
52841         1.587748  1.724797      0             -0.504599   
118474        1.150915 -0.944593      1             -1.501699   

        signup_day_of_week  signup_hour  purchase__hour  category_Ads  \
107357            1.480767    -1.228507       -0.220243             0   
46439            -1.010572     0.651603        1.370816             0   
88685            -0.014036    -0.650012       -0.075601             0   
24733             0.982500    -0.794635        1.226174             1   
96823            -1.010572     0.217732        0.069040             0   
...                    ...          ...             ...           ...   
130664           -0.014036     0.796227       -0.654169             0   
27356            -0.014036    -1.228507       -0.654169             0   
55411            -1.010572    -0.939259        1.660100             0   
52841            -1.010572    -0.360764        1.226174             0   
118474           -1.508839     1.663970       -0.798810             1   

        category_Direct  category_SEO  category_Chrome  category_FireFox  \
107357                0             1                1                 0   
46439                 1             0                0                 0   
88685                 0             1                1                 0   
24733                 0             0                0                 0   
96823                 1             0                0                 0   
...                 ...           ...              ...               ...   
130664                1             0                0                 0   
27356                 1             0                0                 0   
55411                 1             0                0                 1   
52841                 1             0                1                 0   
118474                0             0                1                 0   

        category_IE  category_Opera  category_Safari  category_F  category_M  \
107357            0               0                0           1           0   
46439             0               0                1           1           0   
88685             0               0                0           0           1   
24733             0               0                1           0           1   
96823             1               0                0           0           1   
...             ...             ...              ...         ...         ...   
130664            0               0                1           0           1   
27356             1               0                0           1           0   
55411             0               0                0           1           0   
52841             0               0                0           0           1   
118474            0               0                0           0           1   

        Country_encoded  
107357         0.083790  
46439         -0.036750  
88685          0.083790  
24733         -0.036750  
96823          0.083790  
...                 ...  
130664        -0.501673  
27356         -0.036750  
55411         -0.036750  
52841         -0.036750  
118474         0.083790  

[148848 rows x 18 columns]

In [11]:
import mlflow
mlflow.start_run()

<ActiveRun: >

In [12]:
model.model_decision_tree()

2024/10/27 12:20:34 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:20:34 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/27 12:20:34 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/27 12:20:39 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/27 12:20:40 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:20:40 WAR

New class distribution in training set: [107390  53421]


2024/10/27 12:22:29 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2024/10/27 12:22:33 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detecti

Logging metrics for model: Decision Tree
Dataset name: Fraud, Source: Fraud Data Source

Classification Report:
 {'0': {'precision': 0.9541736737501911, 'recall': 0.9251435982953493, 'f1-score': 0.9394344202148677, 'support': 26985.0}, '1': {'precision': 0.43982251802551303, 'recall': 0.5694793536804309, 'f1-score': 0.49632295415427946, 'support': 2785.0}, 'accuracy': 0.8918710110849849, 'macro avg': {'precision': 0.6969980958878521, 'recall': 0.74731147598789, 'f1-score': 0.7178786871845736, 'support': 29770.0}, 'weighted avg': {'precision': 0.9060558380534082, 'recall': 0.8918710110849849, 'f1-score': 0.8979810969707045, 'support': 29770.0}}
Attempting to save the model...
Attempting to save model of type: <class 'sklearn.tree._classes.DecisionTreeClassifier'>
Model Fraud_Decision_Tree saved at models\Fraud_Decision_Tree.pkl


2024/10/27 12:22:42 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Model save attempted.

Classification Report for Decision Tree:
               precision    recall  f1-score   support

           0       0.95      0.93      0.94     26985
           1       0.44      0.57      0.50      2785

    accuracy                           0.89     29770
   macro avg       0.70      0.75      0.72     29770
weighted avg       0.91      0.89      0.90     29770



Pipeline(steps=[('decision_tree', DecisionTreeClassifier())])

In [13]:
model.model_random_forest()

2024/10/27 12:22:43 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:22:43 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/27 12:22:43 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/27 12:22:47 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/27 12:22:48 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:22:48 WAR

New class distribution in training set: [107390  53421]


2024/10/27 12:25:01 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2024/10/27 12:25:34 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detecti

Logging metrics for model: Random Forest
Dataset name: Fraud, Source: Fraud Data Source

Classification Report:
 {'0': {'precision': 0.954463368860876, 'recall': 0.9988882712618121, 'f1-score': 0.9761706442617608, 'support': 26985.0}, '1': {'precision': 0.9803793328973185, 'recall': 0.5382405745062837, 'f1-score': 0.6949466852109412, 'support': 2785.0}, 'accuracy': 0.9557944239166947, 'macro avg': {'precision': 0.9674213508790972, 'recall': 0.7685644228840479, 'f1-score': 0.835558664736351, 'support': 29770.0}, 'weighted avg': {'precision': 0.9568878216603887, 'recall': 0.9557944239166947, 'f1-score': 0.9498619870243898, 'support': 29770.0}}
Attempting to save random forest model...
Attempting to save model of type: <class 'sklearn.ensemble._forest.RandomForestClassifier'>
Model Fraud_Random_Forest saved at models\Fraud_Random_Forest.pkl


2024/10/27 12:25:44 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Model save attempted.

Classification Report for Random Forest:
               precision    recall  f1-score   support

           0       0.95      1.00      0.98     26985
           1       0.98      0.54      0.69      2785

    accuracy                           0.96     29770
   macro avg       0.97      0.77      0.84     29770
weighted avg       0.96      0.96      0.95     29770



In [14]:
# Other model calls...
model.model_mlp()

2024/10/27 12:25:46 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:25:46 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/27 12:25:46 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/27 12:25:49 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/27 12:25:50 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:25:50 WAR

New class distribution in training set: [107390  53421]
Fitting MLP model...


2024/10/27 12:27:32 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\mlflow\types\utils.py:407: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2024/10/27 12:30:19 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "d:\10 A KAI 2\WEEK 8&9\Fraud_detecti

Logging metrics for model: MLP
Dataset name: Fraud, Source: Fraud Data Source

Classification Report:
 {'0': {'precision': 0.9545520179048445, 'recall': 0.9799147674634056, 'f1-score': 0.9670671274708798, 'support': 26985.0}, '1': {'precision': 0.7379110251450677, 'recall': 0.547935368043088, 'f1-score': 0.6288893467957964, 'support': 2785.0}, 'accuracy': 0.9395028552233793, 'macro avg': {'precision': 0.8462315215249561, 'recall': 0.7639250677532468, 'f1-score': 0.7979782371333382, 'support': 29770.0}, 'weighted avg': {'precision': 0.9342851329590608, 'recall': 0.9395028552233793, 'f1-score': 0.9354304086539463, 'support': 29770.0}}
Attempting to save MLP model...
Attempting to save model of type: <class 'sklearn.neural_network._multilayer_perceptron.MLPClassifier'>
Model Fraud_MLP saved at models\Fraud_MLP.pkl


2024/10/27 12:30:32 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.


Model save attempted.

Classification Report for MLP:
               precision    recall  f1-score   support

           0       0.95      0.98      0.97     26985
           1       0.74      0.55      0.63      2785

    accuracy                           0.94     29770
   macro avg       0.85      0.76      0.80     29770
weighted avg       0.93      0.94      0.94     29770



Pipeline(steps=[('mlp', MLPClassifier(max_iter=1000, random_state=42))])

In [15]:
model.model_lstm()

2024/10/27 12:30:33 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:30:33 WARNING mlflow.sklearn: Failed to infer model signature: the trained model does not have a `predict` or `transform` function, which is required in order to infer the signature
2024/10/27 12:30:33 WARNING mlflow.sklearn: Model was missing function: predict. Not logging python_function flavor!
2024/10/27 12:31:19 WARNING mlflow.models.model: Model logged without a signature and input example. Please set `input_example` parameter when logging the model to auto infer the model signature.
2024/10/27 12:31:20 WARNING mlflow.sklearn: Training metrics will not be recorded because training labels were not specified. To automatically record training metrics, provide training labels as inputs to the model training function.
2024/10/27 12:31:20 WAR

New class distribution in training set: [107390  53421]


d:\10 A KAI 2\WEEK 8&9\Fraud_detection\venv\Lib\site-packages\keras\src\layers\rnn\rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)
2024/10/27 12:33:13 WARNING mlflow.tensorflow: Failed to log training dataset information to MLflow Tracking. Reason: 'Series' object has no attribute 'flatten'


Epoch 1/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6940 - loss: 0.5933

2024/10/27 12:33:58 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 41s 7ms/step - accuracy: 0.6940 - loss: 0.5933
Epoch 2/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.7814 - loss: 0.5040

2024/10/27 12:34:39 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 41s 8ms/step - accuracy: 0.7814 - loss: 0.5039
Epoch 3/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8099 - loss: 0.4646

2024/10/27 12:35:23 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 43s 9ms/step - accuracy: 0.8099 - loss: 0.4646
Epoch 4/50
5023/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8312 - loss: 0.4348

2024/10/27 12:36:51 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 88s 10ms/step - accuracy: 0.8312 - loss: 0.4348
Epoch 5/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8435 - loss: 0.4163

2024/10/27 12:37:39 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 49s 10ms/step - accuracy: 0.8435 - loss: 0.4163
Epoch 6/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8499 - loss: 0.4024

2024/10/27 12:38:33 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 53s 11ms/step - accuracy: 0.8499 - loss: 0.4024
Epoch 7/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8565 - loss: 0.3915

2024/10/27 12:39:50 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 77s 10ms/step - accuracy: 0.8565 - loss: 0.3915
Epoch 8/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8587 - loss: 0.3859

2024/10/27 12:41:09 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 79s 9ms/step - accuracy: 0.8587 - loss: 0.3859
Epoch 9/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8622 - loss: 0.3799

2024/10/27 12:42:24 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 75s 8ms/step - accuracy: 0.8622 - loss: 0.3799
Epoch 10/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8635 - loss: 0.3756

2024/10/27 12:43:05 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 42s 8ms/step - accuracy: 0.8635 - loss: 0.3756
Epoch 11/50
5023/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8680 - loss: 0.3662

2024/10/27 12:43:51 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 46s 9ms/step - accuracy: 0.8680 - loss: 0.3662
Epoch 12/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8689 - loss: 0.3629

2024/10/27 12:45:04 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 72s 7ms/step - accuracy: 0.8689 - loss: 0.3629
Epoch 13/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8635 - loss: 0.3725

2024/10/27 12:45:40 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 36s 7ms/step - accuracy: 0.8635 - loss: 0.3725
Epoch 14/50
5023/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8694 - loss: 0.3596

2024/10/27 12:46:18 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 39s 8ms/step - accuracy: 0.8694 - loss: 0.3596
Epoch 15/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8681 - loss: 0.3614

2024/10/27 12:46:57 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 38s 8ms/step - accuracy: 0.8681 - loss: 0.3614
Epoch 16/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8710 - loss: 0.3565

2024/10/27 12:47:49 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 52s 10ms/step - accuracy: 0.8710 - loss: 0.3565
Epoch 17/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8704 - loss: 0.3567

2024/10/27 12:48:34 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 45s 9ms/step - accuracy: 0.8704 - loss: 0.3567
Epoch 18/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8688 - loss: 0.3584

2024/10/27 12:49:59 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 85s 10ms/step - accuracy: 0.8688 - loss: 0.3584
Epoch 19/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8717 - loss: 0.3539

2024/10/27 12:50:45 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 47s 9ms/step - accuracy: 0.8717 - loss: 0.3539
Epoch 20/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8705 - loss: 0.3533

2024/10/27 12:52:08 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 83s 9ms/step - accuracy: 0.8705 - loss: 0.3533
Epoch 21/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8710 - loss: 0.3516

2024/10/27 12:52:53 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 46s 9ms/step - accuracy: 0.8710 - loss: 0.3516
Epoch 22/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8703 - loss: 0.3515

2024/10/27 12:53:32 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 38s 8ms/step - accuracy: 0.8703 - loss: 0.3515
Epoch 23/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8715 - loss: 0.3488

2024/10/27 12:54:12 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 40s 8ms/step - accuracy: 0.8715 - loss: 0.3488
Epoch 24/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8718 - loss: 0.3477

2024/10/27 12:54:58 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 46s 9ms/step - accuracy: 0.8718 - loss: 0.3477
Epoch 25/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8722 - loss: 0.3464

2024/10/27 12:55:47 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 50s 10ms/step - accuracy: 0.8722 - loss: 0.3464
Epoch 26/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8726 - loss: 0.3465

2024/10/27 12:56:39 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 51s 10ms/step - accuracy: 0.8726 - loss: 0.3465
Epoch 27/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8722 - loss: 0.3453

2024/10/27 12:57:29 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 51s 10ms/step - accuracy: 0.8722 - loss: 0.3453
Epoch 28/50
5023/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8738 - loss: 0.3424

2024/10/27 12:58:27 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 58s 11ms/step - accuracy: 0.8738 - loss: 0.3424
Epoch 29/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8738 - loss: 0.3427

2024/10/27 12:59:49 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 82s 11ms/step - accuracy: 0.8738 - loss: 0.3427
Epoch 30/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.8802 - loss: 0.3340

2024/10/27 13:01:22 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 93s 14ms/step - accuracy: 0.8802 - loss: 0.3340
Epoch 31/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.8819 - loss: 0.3319

2024/10/27 13:02:30 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 68s 13ms/step - accuracy: 0.8819 - loss: 0.3319
Epoch 32/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.8827 - loss: 0.3295

2024/10/27 13:03:31 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 60s 12ms/step - accuracy: 0.8827 - loss: 0.3295
Epoch 33/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.8833 - loss: 0.3289

2024/10/27 13:04:32 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 62s 12ms/step - accuracy: 0.8833 - loss: 0.3289
Epoch 34/50
5021/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8848 - loss: 0.3239

2024/10/27 13:05:29 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 56s 11ms/step - accuracy: 0.8848 - loss: 0.3239
Epoch 35/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8847 - loss: 0.3254

2024/10/27 13:06:25 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 56s 11ms/step - accuracy: 0.8847 - loss: 0.3254
Epoch 36/50
5025/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - accuracy: 0.8877 - loss: 0.3188

2024/10/27 13:07:23 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 59s 12ms/step - accuracy: 0.8877 - loss: 0.3188
Epoch 37/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8866 - loss: 0.3182

2024/10/27 13:08:17 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 53s 10ms/step - accuracy: 0.8866 - loss: 0.3182
Epoch 38/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - accuracy: 0.8859 - loss: 0.3181

2024/10/27 13:10:03 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 107s 21ms/step - accuracy: 0.8859 - loss: 0.3181
Epoch 39/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.8866 - loss: 0.3165

2024/10/27 13:11:29 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 86s 16ms/step - accuracy: 0.8866 - loss: 0.3165
Epoch 40/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.8872 - loss: 0.3162

2024/10/27 13:12:42 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 72s 14ms/step - accuracy: 0.8872 - loss: 0.3162
Epoch 41/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8917 - loss: 0.3075

2024/10/27 13:13:32 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 50s 10ms/step - accuracy: 0.8917 - loss: 0.3075
Epoch 42/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8973 - loss: 0.2976

2024/10/27 13:14:12 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 40s 8ms/step - accuracy: 0.8973 - loss: 0.2976
Epoch 43/50
5022/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8902 - loss: 0.3140

2024/10/27 13:15:04 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 53s 10ms/step - accuracy: 0.8902 - loss: 0.3140
Epoch 44/50
5021/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8975 - loss: 0.2980

2024/10/27 13:15:42 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 38s 8ms/step - accuracy: 0.8975 - loss: 0.2980
Epoch 45/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8964 - loss: 0.2997

2024/10/27 13:16:26 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 44s 9ms/step - accuracy: 0.8964 - loss: 0.2997
Epoch 46/50
5021/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8984 - loss: 0.2953

2024/10/27 13:17:13 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 47s 9ms/step - accuracy: 0.8984 - loss: 0.2953
Epoch 47/50
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.9005 - loss: 0.2906

2024/10/27 13:17:58 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 44s 9ms/step - accuracy: 0.9005 - loss: 0.2906
Epoch 48/50
5023/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8992 - loss: 0.2937

2024/10/27 13:18:43 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 45s 9ms/step - accuracy: 0.8992 - loss: 0.2937
Epoch 49/50
5023/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.9004 - loss: 0.2907

2024/10/27 13:19:25 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 42s 8ms/step - accuracy: 0.9004 - loss: 0.2907
Epoch 50/50
5024/5026 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.8999 - loss: 0.2905

2024/10/27 13:20:07 WARNING mlflow.utils.checkpoint_utils: Checkpoint logging is skipped, because checkpoint 'save_best_only' config is True, it requires to compare the monitored metric value, but the provided monitored metric value is not available.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 41s 8ms/step - accuracy: 0.8999 - loss: 0.2905
1/1 ━━━━━━━━━━━━━━━━━━━━ 3s 3s/step
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 17s 3ms/step
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 16s 3ms/step
5026/5026 ━━━━━━━━━━━━━━━━━━━━ 22s 4ms/step


2024/10/27 13:22:38 WARNING mlflow.sklearn.utils: RocCurveDisplay.from_estimator failed. The artifact training_roc_curve will not be recorded. Artifact error: This Pipeline instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.
2024/10/27 13:22:38 WARNING mlflow.sklearn.utils: PrecisionRecallDisplay.from_estimator failed. The artifact training_precision_recall_curve will not be recorded. Artifact error: This Pipeline instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.


5026/5026 ━━━━━━━━━━━━━━━━━━━━ 20s 4ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step
931/931 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step
Logging metrics for model: LSTM
Dataset name: Fraud, Source: Fraud Data Source

Classification Report:
 {'0': {'precision': 0.954613573110683, 'recall': 0.9914396887159533, 'f1-score': 0.9726781916344004, 'support': 26985.0}, '1': {'precision': 0.867545871559633, 'recall': 0.5432675044883304, 'f1-score': 0.6681386619562817, 'support': 2785.0}, 'accuracy': 0.9495129324823648, 'macro avg': {'precision': 0.911079722335158, 'recall': 0.7673535966021419, 'f1-score': 0.8204084267953411, 'support': 29770.0}, 'weighted avg': {'precision': 0.9464683413733744, 'recall': 0.9495129324823648, 'f1-score': 0.9441883498422081, 'support': 29770.0}}
An error occurred during LSTM model training: Changing param values is not allowed. Param with key='epochs' was already logged with value='50' for run ID='8864f151171443caae4cf353ee9812ed'. Attempted logging new value '20'.

The cause of

Pipeline(steps=[('lstm_classifier',
                 KerasClassifierWrapper(build_fn=<function FraudModel.model_lstm.<locals>.create_lstm_model at 0x0000016ECA356B60>))])

In [16]:
model.display_metrics_table()

Model  Accuracy  Precision    Recall  F1 Score       AUC
0  Decision Tree  0.891871   0.906056  0.891871  0.897981  0.747311
1  Random Forest  0.955794   0.956888  0.955794  0.949862  0.768564
2            MLP  0.939503   0.934285  0.939503  0.935430  0.763925
3           LSTM  0.949513   0.946468  0.949513  0.944188  0.767354

## Model Comparison

| Model            | Accuracy | Precision | Recall   | F1 Score | AUC      |
|------------------|----------|-----------|----------|----------|----------|
| Decision Tree    | 0.891871 | 0.906056  | 0.891871 | 0.897981 | 0.747311 |
| Random Forest    | 0.955794 | 0.956888  | 0.955794 | 0.949862 | 0.768564 |
| MLP              | 0.939503 | 0.934285  | 0.939503 | 0.935430 | 0.763925 |
| LSTM             | 0.949513 | 0.946468  | 0.949513 | 0.944188 | 0.767354 |

### Comparison Analysis

1. **Accuracy**:
   - The **Random Forest** model has the highest accuracy at **95.58%**, indicating that it correctly classifies a majority of instances.
   - The **Decision Tree** model performed the lowest with an accuracy of **89.19%**.

2. **Precision**:
   - The **Random Forest** model also leads in precision (**95.69%**), suggesting it has a lower rate of false positives compared to the other models.
   - The **Decision Tree** model again shows the least precision at **90.61%**.

3. **Recall**:
   - **Random Forest** and **MLP** both achieve high recall values (**95.58%** and **93.95%**, respectively), indicating they successfully identify a high proportion of true positives.
   - The **Decision Tree** model has a recall of **89.19%**, which is lower compared to the other models.

4. **F1 Score**:
   - The **Random Forest** model has the highest F1 score of **94.99%**, which balances precision and recall effectively.
   - The **Decision Tree** again has the lowest F1 score at **89.80%**.

5. **AUC**:
   - The **Random Forest** model also shows the highest AUC at **0.768564**, indicating better performance in distinguishing between classes compared to others.
   - The **Decision Tree** has the lowest AUC at **0.747311**.

### Conclusion

Based on the evaluation metrics, the **Random Forest** model outperforms the other models across all metrics, making it the best choice for this classification task. 

While the **MLP** and **LSTM** models also demonstrate strong performance, they do not surpass the Random Forest model in any of the key metrics. The **Decision Tree** model, while interpretable, shows significantly poorer performance in comparison to the ensemble methods.

In summary, for achieving the best balance between accuracy, precision, recall, and F1 score, the **Random Forest** model is recommended for deployment in practical applications.